pkg/batchscheduler/volcano/volcano_scheduler.go

```diff
@@ -0,0 +1,308 @@
+	"github.com/GoogleCloudPlatform/spark-on-k8s-operator/pkg/util"
+	"github.com/golang/glog"
+	corev1 "k8s.io/api/core/v1"
+	apiextensionsclient "k8s.io/apiextensions-apiserver/pkg/client/clientset/clientset"
+	"k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/resource"
+	"k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/rest"
+
+	"volcano.sh/volcano/pkg/apis/scheduling/v1alpha2"
+	volcanoclient "volcano.sh/volcano/pkg/client/clientset/versioned"
+
+	"github.com/GoogleCloudPlatform/spark-on-k8s-operator/pkg/apis/sparkoperator.k8s.io/v1beta1"
+	"github.com/GoogleCloudPlatform/spark-on-k8s-operator/pkg/batchscheduler/interface"
+)
+
+const (
+	PodGroupName = "podgroups.scheduling.sigs.dev"
+)
+
+type VolcanoBatchScheduler struct {
+	extensionClient apiextensionsclient.Interface
+	volcanoClient   volcanoclient.Interface
+}
+
+func GetPluginName() string {
+	return "volcano"
+}
+
+func (v *VolcanoBatchScheduler) Name() string {
+	return GetPluginName()
+}
+
+func (v *VolcanoBatchScheduler) ShouldSchedule(app *v1beta1.SparkApplication) bool {
+
+	checkScheduler := func(scheduler *string) bool {
+		if scheduler != nil && *scheduler == v.Name() {
+			return true
+		}
+		return false
+	}
+
+	if app.Spec.Mode == v1beta1.ClientMode {
+		return checkScheduler(app.Spec.Executor.SchedulerName)
+	}
+	if app.Spec.Mode == v1beta1.ClusterMode {
+		return checkScheduler(app.Spec.Executor.SchedulerName) && checkScheduler(app.Spec.Driver.SchedulerName)
+	}
+
+	glog.Warningf("Unsupported Spark application mode %s, abandon schedule via volcano.", app.Spec.Mode)
+	return false
+}
+
+func (v *VolcanoBatchScheduler) PatchApplicationPod(pod *corev1.Pod, app *v1beta1.SparkApplication) []util.PatchOperation {
+	var operations []util.PatchOperation
+	if v.ShouldSchedule(app) {
+		// Patch only executor pods
+		if app.Spec.Mode == v1beta1.ClientMode {
+			if util.IsExecutorPod(pod) {
+				patchOperation := v.addVolcanoAnnotation(pod, app)
+				if patchOperation != nil {
+					operations = append(operations, *patchOperation)
+				}
+			}
+			//Patch both driver and executor pods
+		} else if app.Spec.Mode == v1beta1.ClusterMode {
+			if util.IsDriverPod(pod) || util.IsExecutorPod(pod) {
+				patchOperation := v.addVolcanoAnnotation(pod, app)
+				if patchOperation != nil {
+					operations = append(operations, *patchOperation)
+				}
+			}
+		}
+	}
+	return operations
+}
+
+func (v *VolcanoBatchScheduler) addVolcanoAnnotation(pod *corev1.Pod, app *v1beta1.SparkApplication) *util.PatchOperation {
+	path := "/metadata/annotations"
+	var value interface{}
+	if len(pod.Annotations) != 0 {
+		if _, ok := pod.Annotations[v1alpha2.GroupNameAnnotationKey]; ok {
+			return nil
+		}
+		pod.Annotations[v1alpha2.GroupNameAnnotationKey] = v.getAppPodGroupName(app)
+		value = pod.Annotations
+	} else {
+		value = map[string]string{v1alpha2.GroupNameAnnotationKey: v.getAppPodGroupName(app)}
+	}
+	return &util.PatchOperation{Op: "replace", Path: path, Value: value}
```

Comment:
Why `replace`?

Reply:
The annotation has been replaced by a whole collection.